lambdas/indexer/app.py

```diff
@@ -231,95 +225,52 @@ def index(event: chalice.app.SQSEvent):
-        log.info('Recirculating %i token(s), after a delay of %is, for a total value of %i, expiring at %s',
-                 len(tokens), delay, total, datetime.utcfromtimestamp(expiration).isoformat(timespec='seconds') + 'Z')
-        _send_tokens(tokens, delay=delay)
-    elif total < 0:
-        # Discarding token deficit will lead to an overall surplus of token value which will expire at some point.
-        log.info("Discarding token deficit of %i.", total)
+
+    # Consolidate multiple tallies for the same entity and process entities with only one message. Because SQS FIFO
+    # queues try to put as many messages from the same message group in a reception batch, a single message per
+    # group may indicate that that message is the last one in the group. Inversely, multiple messages per group
+    # in a batch are a likely indicator for the presence of even more queued messages in that group. The more
+    # bundle contributions we defer, the higher the amortized savings on aggregation become. Aggregating bundle
+    # contributions is a costly operation for any entity with many contributions e.g., a large project.
+    #
+    tallies_by_id: MutableMapping[EntityReference, List[DocumentTally]] = defaultdict(list)
+    for record in event['Records']:
+        tally = DocumentTally.from_message(record)
+        log.info('Attempt %i of handling %i contribution(s) for entity %s/%s',
+                 tally.attempts, tally.num_contributions, tally.entity.entity_type, tally.entity.entity_id)
+        tallies_by_id[tally.entity].append(tally)
+        # NOTE here we are grouping the tallies based on their entity
+    deferrals, referrals = [], []
+    for tallies in tallies_by_id.values():
+        if len(tallies) == 1:
+            referrals.append(tallies[0])
+        elif len(tallies) > 1:
+            deferrals.append(tallies[0].consolidate(tallies[1:]))
+        else:
+            assert False
+
+    if referrals:
+        for tally in referrals:
+            log.info('Aggregating %i contribution(s) to entity %s/%s',
+                     tally.num_contributions, tally.entity.entity_type, tally.entity.entity_id)
+        indexer_cls = plugin.indexer_class()
+        indexer = indexer_cls()
+        tallies = {tally.entity: tally.num_contributions for tally in referrals}
+        indexer.aggregate(tallies)
+
+    if deferrals:
+        for tally in deferrals:
+            log.info('Deferring aggregation of %i contribution(s) to entity %s/%s',
+                     tally.num_contributions, tally.entity.entity_type, tally.entity.entity_id)
+        document_queue.send_messages(Entries=[dict(tally.to_message(), Id=str(i))
+                                              for i, tally in enumerate(deferrals)])
+
+    document_queue.delete_messages(Entries=[dict(Id=str(i), ReceiptHandle=message.receipt_handle)
+                                            for i, message in enumerate(chain(record))])
```

Comment:
I imagine you meant `event['Records']`, not `record`. Having said that, it's unnecessary to delete the messages from the queue that triggers the lambda. From the [Chalice docs](https://chalice.readthedocs.io/en/latest/topics/events.html#sqs-events):

>If your lambda functions completes without raising an exception, then Lambda will automatically delete all the messages associated with the SQSEvent. You don’t need to manually call sqs.delete_message() in your lambda function.

We had to delete them before since this lambda was triggered by the token queue, It's safe to just delete these two lines.

Reply:
When i was reading through the documentation i opted to add the manual deletion back to the method the reasoning being at the bottom of the documented page you sent...

``` 
Note that if you are using a batch size of more than one, the entire batch succeeds or fails. This means that it is possible for your lambda function to see a message multiple times, even if it’s successfully processed the message previously....
```

I'm not sure if it would be wrong to requeue the whole batch if there was a failure of a single message, or the implications of how that would effect the system overall. If you think I should omit the manual deletion, I'll remove it.

Reply:
I hear your concern about duplicate messages. Deleting a message explicitly will only prevent a successful message from being re-queued if it's done before the failure occurs. Here, we delete all of the messages all at once, explicitly and after _all_ of the aggregation happens. If a failure happens, then all of the messages will be re-queued. (Notice that this was the behaviour before your changes too.)

Fortunately, it doesn't matter. Azul can handle re-processing of a notification. If we wanted to delete messages as they were processed it would be a much more extensive change.

The precedence in the indexer is to not do the explicit delete (look at the `index` lambda for example) so I think you should omit the manual deletion.